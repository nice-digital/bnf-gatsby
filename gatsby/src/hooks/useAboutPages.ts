import { useStaticQuery, graphql } from "gatsby";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface AboutPages {
	allBnfAboutSection: {
		aboutSectionPages: SlugAndTitle[];
	};
	allBnfCautionaryAndAdvisoryGuidance: {
		labelsGuidancePages: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/about/${slug}/`,
});

/**
 * Gets the lists of pages for the about section by combining:
 * - about sections
 * - cautionary/advisory labels
 * - cautionary/advisory guidance page(s)
 * from the feed.
 *
 * @returns The consolidated list of about pages
 */
export const useAboutPages = (): MenuPageLink[] => {
	const { allBnfAboutSection, allBnfCautionaryAndAdvisoryGuidance } =
		useStaticQuery<AboutPages>(
			graphql`
				query AboutPages {
					allBnfAboutSection(sort: { order: ASC }) {
						aboutSectionPages: nodes {
							slug
							title
						}
					}
					allBnfCautionaryAndAdvisoryGuidance {
						labelsGuidancePages: nodes {
							title
							slug
						}
					}
				}
			`
		);

	return [
		...allBnfAboutSection.aboutSectionPages.map(slugToHref),
		{ title: "Labels", href: "/about/labels/" },
		...allBnfCautionaryAndAdvisoryGuidance.labelsGuidancePages.map(slugToHref),
	];
};
