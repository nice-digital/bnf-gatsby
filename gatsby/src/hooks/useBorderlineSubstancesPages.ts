import { useStaticQuery, graphql } from "gatsby";

import { SID, PHPID } from "@nice-digital/gatsby-source-bnf";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface BorderlineSubstancesQueryResult {
	allBnfBorderlineSubstancesTaxonomyRoot: {
		pages: {
			slug: string;
			title: string;
		}[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/borderline-substances/${slug}/`,
});

/**
 * Gets the lists of pages for the borderline substances section
 *
 * @returns The consolidated list of borderline substances pages
 */
export const useBorderlineSubstancesPages = (): MenuPageLink[] => {
	const { allBnfBorderlineSubstancesTaxonomyRoot } =
		useStaticQuery<BorderlineSubstancesQueryResult>(
			graphql`
				{
					allBnfBorderlineSubstancesTaxonomyRoot {
						pages: nodes {
							slug
							title
						}
					}
				}
			`
		);

	return allBnfBorderlineSubstancesTaxonomyRoot.pages.map(slugToHref);
};
