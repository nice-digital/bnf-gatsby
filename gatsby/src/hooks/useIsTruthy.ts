import { useCallback } from "react";

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useIsTruthy = () =>
	useCallback(
		<TMaybe>(maybeT: TMaybe | null): maybeT is TMaybe => !!maybeT,
		[]
	);
