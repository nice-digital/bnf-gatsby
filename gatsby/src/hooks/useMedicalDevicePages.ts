import { useStaticQuery, graphql } from "gatsby";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface MedicalDevicesQueryResult {
	allBnfMedicalDevice: {
		pages: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/medical-devices/${slug}/`,
});

/**
 * Gets the lists of pages for the medical devices section
 *
 * @returns The consolidated list of medical devices pages
 */
export const useMedicalDevicePages = (): MenuPageLink[] => {
	const { allBnfMedicalDevice } = useStaticQuery<MedicalDevicesQueryResult>(
		graphql`
			{
				allBnfMedicalDevice(sort: { fields: title, order: ASC }) {
					pages: nodes {
						slug
						title
					}
				}
			}
		`
	);

	return allBnfMedicalDevice.pages.map(slugToHref);
};
