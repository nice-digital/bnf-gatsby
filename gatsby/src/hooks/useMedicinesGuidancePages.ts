import { useStaticQuery, graphql } from "gatsby";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface MedicinesGuidanceQueryResult {
	allBnfGuidance: {
		pages: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/medicines-guidance/${slug}/`,
});

/**
 * Gets the lists of pages for the medicines guidance section
 *
 * @returns The consolidated list of medicines guidance pages
 */
export const useMedicinesGuidancePages = (): MenuPageLink[] => {
	const { allBnfGuidance } = useStaticQuery<MedicinesGuidanceQueryResult>(
		graphql`
			{
				allBnfGuidance(sort: { fields: order, order: ASC }) {
					pages: nodes {
						slug
						title
					}
				}
			}
		`
	);

	return allBnfGuidance.pages.map(slugToHref);
};
