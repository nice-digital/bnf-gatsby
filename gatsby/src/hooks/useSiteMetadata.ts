import { useStaticQuery, graphql } from "gatsby";

interface SiteMetaData {
	/** Whether the current site is BNF or BNFC */
	isBNF: boolean;
	/** Absolute URL to the site homepage, without trailing slash */
	siteUrl: string;
	/** Short site title e.g. "BNF" or "BNFC" */
	siteTitleShort: string;
	/** Full site title e.g. "British National Formulary" or "British National Formulary for Children" */
	siteTitleLong: string;
}

interface Site {
	site: {
		siteMetadata: SiteMetaData;
	};
}

/**
 * Gets the site metadata, as defined in gatsby-config
 *
 * @returns The site metadata
 */
export const useSiteMetadata = (): SiteMetaData => {
	const { site } = useStaticQuery<Site>(
		graphql`
			query SiteMetaData {
				site {
					siteMetadata {
						isBNF
						siteUrl
						siteTitleShort
						siteTitleLong
					}
				}
			}
		`
	);
	return site.siteMetadata;
};
