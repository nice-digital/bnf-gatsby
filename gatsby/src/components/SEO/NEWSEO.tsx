import { useLocation } from "@reach/router";
import React, { type FC } from "react";

import { useSiteMetadata } from "@/hooks/useSiteMetadata";

export interface NEWSEOProps {
	title?: string;
	description?: string;
	noIndex?: boolean;
	additionalMetadata?: {
		name: string;
		content: string;
	}[];
}

export const NEWSEO: FC<NEWSEOProps> = ({
	title,
	description,
	noIndex,
	additionalMetadata,
}: NEWSEOProps) => {
	const { pathname } = useLocation();
	const { siteUrl, siteTitleShort, siteTitleLong, isBNF } = useSiteMetadata();

	// A global CSS class allows us to distinguish BNF from BNFC in (S)CSS styles
	const globalCSSClassName = isBNF ? "site-bnf" : "site-bnfc";

	const defaultDescription = `Online access to the ${siteTitleShort} (${siteTitleLong}) content within the UK, provided by NICE`;

	return (
		<>
			{/* <Helmet
				title={title}
				titleTemplate={`%s | ${siteTitleShort} | NICE`}
				defaultTitle={`${siteTitleShort} (${siteTitleLong}) | NICE`}
			> */}

			<title>
				{title
					? `${title} | ${siteTitleShort} | NICE`
					: `${siteTitleShort} (${siteTitleLong}) | NICE`}
			</title>
			<html lang="en-GB" className={globalCSSClassName} />
			<meta name="description" content={description || defaultDescription} />
			<meta
				property="og:description"
				content={description || defaultDescription}
			/>
			<meta property="og:url" content={siteUrl + pathname} />
			<meta property="og:locale" content="en_GB" />
			<meta property="og:type" content="website" />
			<meta
				property="og:title"
				content={
					(title ? `${title} | ` : "") +
					`${siteTitleShort} content published by NICE`
				}
			/>
			<meta name="twitter:card" content="summary_large_image" />
			<meta name="twitter:site" content="@NICEcomms" />
			<meta name="twitter:creator" content="@NICEcomms" />
			<meta name="theme-color" content={isBNF ? "#93da49" : "#ef4a81"} />
			{additionalMetadata?.map((x, i) => (
				<meta key={i} name={x.name} content={x.content} />
			))}
			{noIndex ? <meta name="robots" content="noindex" /> : null}
			<link rel="icon" href="/favicon.ico" />
			<link rel="canonical" href={siteUrl + pathname} />
			{/* </Helmet> */}
		</>
	);
};
