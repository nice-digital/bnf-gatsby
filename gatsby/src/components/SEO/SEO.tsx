import { useLocation } from "@reach/router";
import React, { type FC } from "react";
import { Helmet } from "react-helmet";

import { useSiteMetadata } from "@/hooks/useSiteMetadata";

export interface SEOProps {
	title?: string;
	description?: string;
	noIndex?: boolean;
	additionalMetadata?: {
		name: string;
		content: string;
	}[];
}

export const SEO: FC<SEOProps> = ({
	title,
	description,
	noIndex,
	additionalMetadata,
}: SEOProps) => {
	const { pathname } = useLocation();
	const { siteUrl, siteTitleShort, siteTitleLong, isBNF } = useSiteMetadata();

	// A global CSS class allows us to distinguish BNF from BNFC in (S)CSS styles
	const globalCSSClassName = isBNF ? "site-bnf" : "site-bnfc";

	const defaultDescription = `Free online access to the UK ${siteTitleShort} (${siteTitleLong}) content published by NICE`;

	return (
		<>
			<Helmet
				title={title}
				titleTemplate={`%s | ${siteTitleShort} content published by NICE`}
				defaultTitle={`${siteTitleShort} (${siteTitleLong}) | NICE`}
			>
				<html lang="en-GB" className={globalCSSClassName} />
				<meta name="description" content={description || defaultDescription} />
				<meta
					property="og:description"
					content={description || defaultDescription}
				/>
				<meta property="og:url" content={siteUrl + pathname} />
				<meta property="og:locale" content="en_GB" />
				<meta property="og:type" content="website" />
				<meta
					property="og:title"
					content={
						(title ? `${title} | ` : "") +
						`${siteTitleShort} content published by NICE`
					}
				/>
				<meta property="og:image" content={`${siteUrl}/open-graph-image.png`} />
				<meta property="og:image:width" content="1200" />
				<meta property="og:image:height" content="630" />
				<meta name="twitter:card" content="summary" />
				<meta
					name="twitter:image"
					content={`${siteUrl}/twitter-summary-image.png`}
				/>
				<meta
					name="twitter:image:alt"
					content={`NICE and ${siteTitleShort} logos`}
				/>
				<meta name="twitter:site" content="@NICEcomms" />
				<meta name="twitter:creator" content="@NICEcomms" />
				<meta name="theme-color" content="#004650" />
				{additionalMetadata?.map((x, i) => (
					<meta key={i} name={x.name} content={x.content} />
				))}
				{noIndex ? <meta name="robots" content="noindex" /> : null}
				<link rel="icon" href="/favicon.ico" />
				<link rel="canonical" href={siteUrl + pathname} />
			</Helmet>
		</>
	);
};
