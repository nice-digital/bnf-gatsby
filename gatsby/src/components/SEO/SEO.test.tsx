/* eslint-disable testing-library/no-node-access */
import { render, waitFor } from "@testing-library/react";
import React from "react";

import { useSiteMetadata } from "@/hooks/useSiteMetadata";

import { SEO } from "./SEO";

describe("SEO", () => {
	it("should set HTML language to UK English", async () => {
		render(<SEO title="Custom title" description="Custom description" />);
		await waitFor(() => {
			expect(document.querySelector("html")?.getAttribute("lang")).toEqual(
				"en-GB"
			);
		});
	});

	it("should add BNF specific CSS class to html root element", async () => {
		render(<SEO />);
		await waitFor(() => {
			expect(document.documentElement).toHaveClass("site-bnf");
		});
	});

	it("should add BNFC specific CSS class to html root element", async () => {
		(useSiteMetadata as jest.Mock).mockImplementationOnce(() => ({
			isBNF: false,
		}));
		render(<SEO />);
		await waitFor(() => {
			expect(document.documentElement).toHaveClass("site-bnfc");
		});
	});

	it("should set default title and description when none provided", async () => {
		render(<SEO />);

		await waitFor(() => {
			expect(
				document
					.querySelector("meta[property='og:title']")
					?.getAttribute("content")
			).not.toBeUndefined();
		});

		expect(document.querySelector("head")).toMatchSnapshot();
	});

	it("should render custom title and description in document head", async () => {
		render(<SEO title="Custom title" description="Custom description" />);
		await waitFor(() => {
			expect(
				document
					.querySelector("meta[property='og:title']")
					?.getAttribute("content")
			).toEqual("Custom title | BNF content published by NICE");
		});

		expect(document.title).toEqual(
			"Custom title | BNF content published by NICE"
		);

		expect(
			document
				.querySelector("meta[property='og:title']")
				?.getAttribute("content")
		).toEqual("Custom title | BNF content published by NICE");
		expect(
			document
				.querySelector("meta[name='description']")
				?.getAttribute("content")
		).toEqual("Custom description");
		expect(
			document
				.querySelector("meta[property='og:description']")
				?.getAttribute("content")
		).toEqual("Custom description");
	});

	it("should render robots when noIndex is set to true", async () => {
		render(<SEO noIndex={true} />);
		await waitFor(() => {
			expect(document.querySelector("meta[name='robots']")).toHaveAttribute(
				"content",
				"noindex"
			);
		});
	});

	it("should not render robots when noIndex is set to false", async () => {
		render(<SEO noIndex={false} />);
		await waitFor(() => {
			expect(document.querySelector("meta[name='robots']")).toBeNull();
		});
	});

	it("should not render robots when noIndex is not set", async () => {
		render(<SEO />);
		await waitFor(() => {
			expect(document.querySelector("meta[name='robots']")).toBeNull();
		});
	});
});
