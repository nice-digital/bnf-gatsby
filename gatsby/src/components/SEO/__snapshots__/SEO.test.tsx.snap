// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SEO should set default title and description when none provided 1`] = `
<head>
  <title>
    BNF (British National Formulary) | NICE
  </title>
  <link
    data-react-helmet="true"
    href="/favicon.ico"
    rel="icon"
  />
  <link
    data-react-helmet="true"
    href="https://bnf.nice.org.uk/test/"
    rel="canonical"
  />
  <meta
    content="Online access to the BNF (British National Formulary) content within the UK, provided by NICE"
    data-react-helmet="true"
    name="description"
  />
  <meta
    content="Online access to the BNF (British National Formulary) content within the UK, provided by NICE"
    data-react-helmet="true"
    property="og:description"
  />
  <meta
    content="https://bnf.nice.org.uk/test/"
    data-react-helmet="true"
    property="og:url"
  />
  <meta
    content="en_GB"
    data-react-helmet="true"
    property="og:locale"
  />
  <meta
    content="website"
    data-react-helmet="true"
    property="og:type"
  />
  <meta
    content="BNF content published by NICE"
    data-react-helmet="true"
    property="og:title"
  />
  <meta
    content="summary_large_image"
    data-react-helmet="true"
    name="twitter:card"
  />
  <meta
    content="@NICEcomms"
    data-react-helmet="true"
    name="twitter:site"
  />
  <meta
    content="@NICEcomms"
    data-react-helmet="true"
    name="twitter:creator"
  />
  <meta
    content="#93da49"
    data-react-helmet="true"
    name="theme-color"
  />
</head>
`;
